### ISSUE TITLE
Child records not mapping ParentId even when the Id is on the object

### ISSUE THREAD

**Issue by elexisvenator** [jonwagner/Insight.Database#238]
So I have a rather simple query, which is almost identical to the one on the [Specifying Result Structures](https://github.com/jonwagner/Insight.Database/wiki/Specifying-Result-Structures) page:

```
CREATE PROC GetBeerAndGlasses AS
    SELECT * FROM Beer
    SELECT b.BeerID, g.* FROM Beer JOIN Glasses ON (...)

class Glass
{
    public int ID;
    public int BeerID;
}

class Beer
{
    public int ID;
    public IList<Glass> Glasses;
}

var results = connection.Query("GetBeerAndGlasses", Parameters.Empty,
    Query.Returns(Some<Beer>.Records)
        .ThenChildren(Some<Glasses>.Records);
```

The key difference from the wiki example is that `Glass` also has `BeerID` as a property. For our use, this is important because all Glasses belong to exactly one Beer, and we often look at Glass on its own without its matching Beer, so the object has BeerID on it. In this example BeerID never gets mapped, I would guess because Insight uses it for mapping to the parent object then discards the column.

Reproduced in SqlServer 2014 Standard, Insight 5.2.5

**Comment by kevinsea** [comment]
Mapping happens one of several way, by query hints, attributes or by convention.  Avoid mapping by convention when you use just 'ID' for the PK of your object, it will not set BeerId in the Glass object.

1) I always use repository dto's, so I don't mind the attributes on my classes:  

class Glass
{
   [RecordId] public int ID;
   [ParentRecordId] public int BeerID;
}

class Beer
{
    [RecordId] public int ID;
    public IList<Glass> Glasses;
}

2) This would auto map just fine:

class Glass
{
    public int GlassID;
    public int BeerID;
}

class Beer
{
    public int BeerID;
    public IList<Glass> Glasses;
}

3) The hint example from the page you referenced:

var results = connection.Query("GetBeerAndPossibleGlasses", Parameters.Empty,
    Query.Returns(Some<Beer>.Records)
        .ThenChildren(
            Some<Glasses>.Records,
            id: beer => beer.ID,
            into: (beer, glass) => beer.Glasses = glass);

**Comment by elexisvenator** [comment]
The workaround we have now is to do the following:

```
var results = connection.Query("GetBeerAndGlasses", Parameters.Empty,
    Query.Returns(Some<Beer>.Records)
        .ThenChildren(Some<Glasses>.Records.GroupBy(glass => glass.Id));
```

This works, and your suggestions work as well, but:

 * I am working on a large codebase with an established convention for defining these kind of objects.
 * We do not like using ORM-specific attributes on objects that otherwise don't care about how they are accessed.
 * For Insight, it feels inconsistent for the Id to be mapped in almost every scenario.. except the simplest.

**Comment by kevinsea** [comment]
The GroupBy is definitely more readable than the 3rd option I gave you, I agree its the cleanest solution for you right now.

Issue #182 covers the issue you are experiencing.

[Specifying-Result-Structures](https://github.com/jonwagner/Insight.Database/wiki/Specifying-Result-Structures) provides all of the mapping rules: Your GroupBy is rule 1 and before I think you were relying on the last rule.  Jon talks about how that last rule works a bit differently in 182 as well. 

The short of those links is that auto mapping is a bit more complicated than it seems.  Though you have a reasonable expectation that child property child.BeerId should map to Beer.Id.  I'm happy to look at this a bit and see if there is a narrow fix to the automapping logic

**Comment by elexisvenator** [comment]
I never realised that by default it maps by the last rule. I thought this situation would have been picked up by rule 3: *A field called "ParentID"*. Does rule 3 look for the literal name "ParentId" instead of "{name-of-parent-class}ID" (in this example that would be BeerId)?

**Comment by kevinsea** [comment]
Yeah, that's literally 'ParentID' in that rule.  We'd ideally also support <ParentClassName>ID and <ParentClassName>_ID too.

**Comment by kevinsea** [comment]
Created pull request https://github.com/jonwagner/Insight.Database/pull/240 for Jon's review.

**Comment by jonwagner** [comment]
I'll look at the PR

**Comment by jonwagner** [comment]
Fixed in v5.2.8
